src/providers/moduleProvider.ts

```diff
@@ -0,0 +1,161 @@
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { ExecuteCommandParams, ExecuteCommandRequest } from 'vscode-languageclient';
+import { Utils } from 'vscode-uri';
+import { ClientHandler } from '../clientHandler';
+
+const LOCALMODULE = new vscode.ThemeIcon('symbol-folder', new vscode.ThemeColor('terminal.ansiBrightBlue'));
+const TFREGISTRY = new vscode.ThemeIcon('extensions-view-icon', new vscode.ThemeColor('terminal.ansiBrightMagenta'));
+const GITHUBMODULE = new vscode.ThemeIcon('github');
+
+class TerraformModule extends vscode.TreeItem {
+  constructor(
+    public label: string,
+    public provider: string,
+    public version: string,
+    public type: string,
+    public doclink: string,
+    public state: vscode.TreeItemCollapsibleState,
+    public readonly children?: [TerraformModule],
+  ) {
+    super(label, state);
+    if (this.version != '') {
+      this.tooltip = `${this.provider}@${this.version}`;
+      this.description = `${this.provider}@${this.version}`;
+    } else {
+      this.tooltip = `${this.provider}`;
+      this.description = `${this.provider}`;
+    }
+  }
+
+  iconPath = this.getIcon(this.type);
+
+  getIcon(type: string) {
+    const icon = this.terraformIcon();
+    switch (type) {
+      case 'tfregistry':
+        return icon;
+      case 'local':
+        return LOCALMODULE;
+      case 'github':
+        return GITHUBMODULE;
+      default:
+        return TFREGISTRY;
+    }
+  }
+
+  private terraformIcon() {
+    // need current extension path to find icon svg
+    // could possibly make this a custom icon
+    const myExtDir = vscode.extensions.getExtension('hashicorp.terraform').extensionPath;
```

Comment:
No, but it was a quick and low cost method to get the path. Using `absolutePath()` requires passing in the `vscode.ExtensionContext` into the TreeProvider and then into the constructor of each TreeItem. Will examine similar approaches and think on this.

Reply:
Modified to pass down the context and build path off that. Still not happy with the constructor for these, but I think this is the best solution at the moment.

Reply:
You could pass just the path to the parent directory instead of the whole context I suppose?

It's largely the testability of any code which calls here that was my concern though, as explained in https://github.com/hashicorp/vscode-terraform/issues/786

Reply:
I could pass a `vscode.Uri` from `context.extensionUri` to the constructor, but I also pass the context in order to register the commands. I prefer to keep command registration and related event handlers close to the calling code, but they could be moved to `extenion.ts` if we wanted to remove the context entirely from the class. 

I do agree that mocking vscode.ExtensionContext is difficult, but so will mocking `ClientHandler` be, as it has alot of settings and other logic inside it. I can't remove `ClientHandler` because that's how I call the LS.

Reply:
Hmm, it's tightly coupled either way I guess we have no easy way out of it 🤔